#  Exploring Used cars from eBay 

## 1. Introduction 

eBay is an important website where basically sellers create listings for their items and they include the item, photos, price payment and others. Also, there are auctions , here the highest bidder wins the item.

## 2. Aim of the project
The objective of this project is to work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website and clean data and analyze the used car listings. The original dataset can be found [here](https://data.world/data-society/used-cars-data).

In [1]:
import pandas as pd
import numpy as np

In [2]:
#reading data into pandas with encoding
autos=pd.read_csv("autos.csv",encoding="Latin-1")

In [3]:
#fist view at data
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

## First observations

After checking few rows and info method we can observe there are 50000 rows × 20 columns  most of which are strings, there are int64(5), object(15) and few null values in columns and also this data frame contains some basic data for each used car such as: name of the car, seller, price model which are detailed as follows:

* **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
* **name** - Name of the car.
* **seller** - Whether the seller is private or a dealer.
* **offerType** - The type of listing
* **price** - The price on the ad to sell the car.
* **abtest** - Whether the listing is included in an A/B test.
* **vehicleType** - The vehicle Type.
* **yearOfRegistration** - The year in which the car was first registered.
* **gearbox** - The transmission type.
* **powerPS** - The power of the car in PS.
* **model** - The car model name.
* **kilometer** - How many kilometers the car has driven.
* **monthOfRegistration** - The month in which the car was first registered.
* **fuelType** - What type of fuel the car uses.
* **brand** - The brand of the car.
* **notRepairedDamage** - If the car has a damage which is not yet repaired.
* **dateCreated** - The date on which the eBay listing was created.
* **nrOfPictures** - The number of pictures in the ad.
* **postalCode** - The postal code for the location of the vehicle.
* **lastSeenOnline** - When the crawler saw this ad last online.

## 3. Renaming columns

Now let's revising some columns names we can find some string structure as camelcase and we will turn it to snakecase. For instance:
* yearOfRegistration ** to ** registration_year

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
#save the changes as dictionary
new_col={ 
    "yearOfRegistration":"registration_year",
    "monthOfRegistration":"registration_month",
    "notRepairedDamage":"unrepaired_damage",
    "dateCreated":"ad_created",
    "dateCrawled":"date_crawled",
    "offerType":"offer_type",
    "vehicleType":"vehicle_type",
    "powerPS":"power_ps",
    "fuelType":"fuel_type",
    "nrOfPictures":"nr_of_picture",
    "postalCode":"postal_code",
    "lastSeen":"last_seen"
}

In [7]:
#replacing old columns with new columns
autos.rename(new_col,axis=1,inplace=True)

In [8]:
#checking new columns name
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_picture  \
0     peugeot              nein  2016-03-26 00:00:00              0   
1         bmw              nein  2016-04-04 00:00:00              0   
2  volkswagen              nein  2016-03-26 00:00:00              0   
3       smart              nein  2016-03-12 00:00:00              0   
4        ford              nein  2016-04-01 00:00:00              0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

After changed the structure of columns name as the previous step changing camelcase to snakecase we could handle easier the data frame and they are more descriptive. 

## 4. Removing not important columns

Now let's do some basic data exploration to determine what other cleaning tasks need to be done

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_picture   postal_code            last_seen  
count                 50000        50000.0  50000.000000                50000  
unique                   76            NaN           NaN                39481  
top     2016-04-03 00:00:00            NaN           NaN  2016-04-07 06:17:27  
freq                   1946            NaN           NaN                    8  
mean                    NaN            0.0  50813.627300                  NaN  
std                     NaN            0.0  25779.747957                  NaN  
min                     NaN            0.0   1067.000000                  NaN  
25%                     NaN            0.0  30451.000000                  NaN  
50%                     NaN            0.0  49577.000000                  NaN  
75%                     NaN            0.0  71540.000000                  NaN  
max                     NaN            0.0  99998.000000                  NaN

* Here we can find "price" and "odometer" columns are numeric values but they are stored as text.
* There is a number of text columns where almost all of the values are the same (seller and offer_type).
* column number of photos do not contribute to our analysis

In [10]:
autos.drop(columns=['seller','offer_type','nr_of_picture'], inplace=True)

## 5. Turning text columns into numeric columns 

In [11]:
#checking type
print(autos["price"].dtype)
print(autos["odometer"].dtype)

object
object


In [12]:
#checking unique values for both columns
print(autos["price"].unique())

print("\n")

print(autos["odometer"].unique())

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


There are some values that make these two values as text "$" for price column and "km" for odometer

In [13]:
#replacing $ 
autos["price"]=autos["price"].str.replace("$","")
print(autos["price"].unique())

['5,000' '8,500' '8,990' ... '385' '22,200' '16,995']


In [14]:
#replacing km 
autos["odometer"]=autos["odometer"].str.replace("km","")
print(autos["odometer"].unique())

['150,000' '70,000' '50,000' '80,000' '10,000' '30,000' '125,000' '90,000'
 '20,000' '60,000' '5,000' '100,000' '40,000']


However, we need to replace "," to " " in order to convert them to numeric dtype

In [15]:
autos["price"]=autos["price"].str.replace(",","")
print(autos["price"].unique())

['5000' '8500' '8990' ... '385' '22200' '16995']


In [16]:
autos["odometer"]=autos["odometer"].str.replace(",","")
print(autos["odometer"].unique())

['150000' '70000' '50000' '80000' '10000' '30000' '125000' '90000' '20000'
 '60000' '5000' '100000' '40000']


Now we can convert these two columns to numeric dtype

In [17]:
autos["odometer"]=autos["odometer"].astype(int)
print(autos["odometer"].unique())

[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]


In [18]:
autos["price"]=autos["price"].astype(int)
print(autos["price"].unique())

[ 5000  8500  8990 ...   385 22200 16995]


## 5.1 renaming columns

So let´s rename these two columns as folllows:

 odometer to dometer_km   
 
 and
 
 price to price_$

In [19]:
#renaming columns name for price and odometer
autos.rename({"price":"price_$"},axis=1,inplace=True)

In [20]:
autos.rename({"odometer":"odometer_km"},axis=1,inplace=True)

## 6. Deleting outlier values

Let´s continue exploring this data, especially focus on data that does not look right starting by "odomter_km" and "price_$" columns:
we will analyze them using minimum and maximun values and look for any values that look too low or too high  ( outliers) and proceed to remove them.

In [21]:
#there are 13 unique values for this column
autos["odometer_km"].unique().shape

(13,)

In [22]:
#there are 2357 unique values for this column
autos["price_$"].unique().shape

(2357,)

## 6.1 Odometer column

In [23]:
#there no outliers data for this column
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [24]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

After checking min and max data for odometer column we can conclude that all information is right and not outliers are present.

## 6.2 Price column

In [25]:
#cars price in ascending order
autos["price_$"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_$, dtype: int64

There are really low prices even 0 dollars and 1 dollar, maybe this cars were not set with the real price since seller want to negotiate the price, we will remove them.

In [26]:
autos["price_$"].value_counts().sort_index(ascending=False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price_$, dtype: int64

As we can see there are some prices that are too high, prices from 99999999 to 999990 have 15 entries. let's proceed to remove them

In [27]:
#removing outliers
autos=autos[autos["price_$"].between(2,999989)]

In [28]:
autos["price_$"].describe()

count     48409.000000
mean       5907.909707
std        9068.263463
min           2.000000
25%        1250.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

In [29]:
autos["price_$"].value_counts().sort_index(ascending=True).head(10)

2     3
3     1
5     2
8     1
9     1
10    7
11    2
12    3
13    2
14    1
Name: price_$, dtype: int64

After removing outlier values we pass from 50.000 entries to 48.409 entries and data is more reliable. min price 2 dollars  and  max 350,000 dollars

## 7. Date distribution

Keeping analyzising and cleanning data let's move on the date columns and understand the date range.

There are 5 columns that should represent date values,the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. We need to convert data for these three columns into a numerical representation to undertand it quantitatively. The other two columns are represented as numerics values and we can undertand the distribution without any extra data processing.

For the three columns prevously mentioned we can found that they are represented as full timestamp values

In [30]:
autos[['date_crawled','last_seen','ad_created']][0:5]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00

The first 10 characters represent the year,month and day.
Now we can analyse the date distribution for each of these three columns as percentage.

In [31]:
autos["date_crawled"].str[:7].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03    0.838088
2016-04    0.161912
Name: date_crawled, dtype: float64

In [32]:
autos["ad_created"].str[:7].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06    0.000021
2015-08    0.000021
2015-09    0.000021
2015-11    0.000021
2015-12    0.000041
2016-01    0.000248
2016-02    0.001260
2016-03    0.837510
2016-04    0.160859
Name: ad_created, dtype: float64

In [33]:
autos["last_seen"].str[:7].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03    0.424033
2016-04    0.575967
Name: last_seen, dtype: float64

After run the previous cells we can say that the ads were first crawled only until march 2016 but the first add were created on june, this took approximately 9 months and the last seen is concretated in the first days of april. The columns last_Seen and date_Crawled could be used to estimate how long a car will be at least online before it is sold, based on this date crawled is grouped on march and the last seen is concentrated on first days on April, so took at least less of one month to be sold from the moment of the this ad was first crawled.

Following we will look at descriptive statistics of registration year column

In [34]:
autos["registration_year"].describe()

count    48409.000000
mean      2004.774319
std         88.783278
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We can found that there some odd values as follows:
* The minimum value is 1000, before cars were invented
* The maximum value is 9999, many years into the future

Because a car can´t be first registered after the listing was seen, any vehicle with a registration year above 2016 is inaccurate, determinate the earliest valid year is more diffcult.
Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need othe parameter.

In [35]:
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000083
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000041
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
1955    0.000041
1956    0.000083
1957    0.000041
1958    0.000083
1959    0.000124
1960    0.000454
1961    0.000124
1962    0.000083
1963    0.000165
1964    0.000248
          ...   
2000    0.064616
2001    0.054370
2002    0.051230
2003    0.055713
2004    0.055775
2005    0.060464
2006    0.055155
2007    0.046975
2008    0.045715
2009    0.043029
2010    0.032824
2011    0.033527
2012    0.027061
2013    0.016567
2014    0.013675
2015    0.008077
2016    0.024975
2017    0.028631
2018    0.009709
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000083
5911    0.000021
6200    0.000021
8888    0.0000

As we can see the proportion of the total in percentage is mostly between 1910-2016, we will discard the rest of values.

In [36]:
autos=autos[autos["registration_year"].between(1900,2016)]

In [37]:
autos["registration_year"].describe()

count    46531.000000
mean      2002.926049
std          7.159928
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

After eliminate values outside 1910-2016 values we can say that the most of cars have registration of 2008 followed by 2003 year.

## 8. Identifying brands and prices

Now we will check brands of this dataset.
First let's figure out brands and sells for each one

In [38]:
autos["brand"].value_counts()

volkswagen        9834
bmw               5116
opel              4996
mercedes_benz     4488
audi              4030
ford              3256
renault           2193
peugeot           1390
fiat              1195
seat               850
skoda              764
nissan             712
mazda              709
smart              661
citroen            653
toyota             593
hyundai            467
sonstige_autos     449
volvo              425
mini               409
mitsubishi         381
honda              365
kia                330
alfa_romeo         309
porsche            284
suzuki             276
chevrolet          265
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru              99
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

We will only take for this analysis top 10 brands and then assign them to a dictionary with mean price for each brand.

In [39]:
autos["brand"].value_counts().head(10)

volkswagen       9834
bmw              5116
opel             4996
mercedes_benz    4488
audi             4030
ford             3256
renault          2193
peugeot          1390
fiat             1195
seat              850
Name: brand, dtype: int64

In [40]:
top_brands = {}
for b in autos['brand'].value_counts().index[:10]:
    select = autos[autos['brand'] == b]
    mean_price = int(select['price_$'].mean())
    top_brands[b] = mean_price

In [41]:
top_brands

{'audi': 9362,
 'bmw': 8367,
 'fiat': 2818,
 'ford': 3757,
 'mercedes_benz': 8657,
 'opel': 2990,
 'peugeot': 3100,
 'renault': 2483,
 'seat': 4412,
 'volkswagen': 5417}

The top 10 sellest brand we can find the next:
* Audi,Bmw and Mercedes are the most expensive cars 
* the less expensive are fiat,opel and renault
* volskwagen is the sellest car and it is in between

## 9. Finding the mean mileage

It is time to find the mean mileage and aggregate it to a table together to brand and mean price for this we will find mean mileage as the previous step and then created a dataframe with these data in order to display it and like this will be more comfortable to see.

First let,s repeat the loop to find mean mileage

In [42]:
avg_mileage = {}
for b in autos['brand'].value_counts().index[:10]:
    select = autos[autos['brand'] == b]
    mean_price = int(select['odometer_km'].mean())
    avg_mileage[b] = mean_price

In [43]:
avg_mileage

{'audi': 129208,
 'bmw': 132553,
 'fiat': 117066,
 'ford': 124210,
 'mercedes_benz': 130838,
 'opel': 129322,
 'peugeot': 127122,
 'renault': 128144,
 'seat': 121029,
 'volkswagen': 128708}

Then we will convert both dictonaries into series objects

In [44]:
price_series=pd.Series(top_brands)
price_series

audi             9362
bmw              8367
fiat             2818
ford             3757
mercedes_benz    8657
opel             2990
peugeot          3100
renault          2483
seat             4412
volkswagen       5417
dtype: int64

In [45]:
mileage_series=pd.Series(avg_mileage)
mileage_series

audi             129208
bmw              132553
fiat             117066
ford             124210
mercedes_benz    130838
opel             129322
peugeot          127122
renault          128144
seat             121029
volkswagen       128708
dtype: int64

Now we will create a dataframe from the first series object using the dataframe constructor

In [46]:
top_brands=pd.DataFrame(price_series,columns=["mean_price"])
top_brands

mean_price
audi                 9362
bmw                  8367
fiat                 2818
ford                 3757
mercedes_benz        8657
opel                 2990
peugeot              3100
renault              2483
seat                 4412
volkswagen           5417

We will check the mean price of the top 10

In [47]:
top_brands.describe()

mean_price
count    10.000000
mean   5136.300000
std    2674.214321
min    2483.000000
25%    3017.500000
50%    4084.500000
75%    7629.500000
max    9362.000000

Done the previous step we are going to add mileage series to this new data frame

In [48]:
top_brands["mean_mileage"]=mileage_series

Also let´s add quantities sold for each brand toour new data frame

In [49]:
q_sold=autos["brand"].value_counts().head(10)

In [50]:
quantities_sold=pd.Series(q_sold)
quantities_sold

volkswagen       9834
bmw              5116
opel             4996
mercedes_benz    4488
audi             4030
ford             3256
renault          2193
peugeot          1390
fiat             1195
seat              850
Name: brand, dtype: int64

In [51]:
top_brands["quantities_sold"]=quantities_sold

In [52]:
top_brands

mean_price  mean_mileage  quantities_sold
audi                 9362        129208             4030
bmw                  8367        132553             5116
fiat                 2818        117066             1195
ford                 3757        124210             3256
mercedes_benz        8657        130838             4488
opel                 2990        129322             4996
peugeot              3100        127122             1390
renault              2483        128144             2193
seat                 4412        121029              850
volkswagen           5417        128708             9834

Now we have dataframe complete we can compare brand, mean price mean mileage adn quantities sold. 
the TOP 10 brands show that mean mileage is between 121,000 to 132,000 Km so all these used cars have not a big difference in mileage.
we can observe that volkswagen is the sellest car by far and its price is between the most expensives ones and the less expensives ones as well as mena mileage into this top. 
Audi is the most expensive brand in this top adn their sells is in the middle of the top.
Renault is the cheapest in this top but sells do not make a difference.

## 10. Analyzing Volkswagen

For this last step of this project we will figure out the sellest model of the sellest brand, Volkswagen.

In [53]:
volkswagen=autos[autos["brand"]=="volkswagen"]

In [54]:
volkswagen["model"].value_counts()

golf           3697
polo           1603
passat         1346
transporter     674
touran          433
lupo            321
sharan          220
caddy           203
beetle          123
tiguan          118
bora            100
andere           96
touareg          94
scirocco         85
fox              82
eos              66
kaefer           56
up               51
jetta            38
phaeton          31
cc               18
amarok            6
Name: model, dtype: int64

Let's check mean price for the sellest model'

In [55]:
avg_price_v = {}
for m in volkswagen["model"].value_counts().index[:5]:
    select = autos[autos['model'] == m]
    mean_price = int(select['price_$'].mean())
    avg_price_v[m] = mean_price

In [56]:
avg_price_v

{'golf': 5095,
 'passat': 4959,
 'polo': 2603,
 'touran': 7914,
 'transporter': 10374}

The sellest model by far is Golf with 3697 units followed by polo with 1603 units. Golf is the model most search in ebay Germany and its price is a little bit under the mean price of all used cars list and also it is in the same situation for mean of the mean price of top 10 brands.

## 11. CONCLUSIONS

After data cleaning and finalized our analysis we can say:
* Some values are odd or outlier as price and year registration for the analysis, with the cleaning we depured data and worked with more reliable information.
* Took least of 1 month sell a car in ebay germany in this analysis 
* the top 10 brands, 5 of them are from germany so germans usually buy national brands by ebay.
* the 5 most sellest brands are from Germany Volkswagen,BMW,Opel, Mercedes Benz and Audi. 
* Volkswagen was the sellest brand by far with respect to the second place of the top 10 sellest brands
* The price do not influenced in the purchase of brand or other since there are other brand cheaper than Volkswagen.
* the Golf model of Volkswagen sold 3697 units in this analysis followed by Polo, 1603 units. it indicates that Golf sold more units than 3 brands of the top 10 like as fiat,peugeot and seat.